users/test.py

```diff
@@ -36,36 +40,139 @@ def setUp(self):
+
+        Note: app.upsert_user_in_db should not insert the user if it is missing
+        the 'user_id' or 'name' attributes, but should insert the user if
+        the user has more attributes than those. If the user is malformatted,
+        app.upsert_user_in_db raises an AttributeError, else it upserts the
+        user and returns the new user object.
+        """
+        self.mock_collection = mongomock.MongoClient().db.collection
+        # missing name
+        self.assertRaises(AttributeError, app.upsert_user_in_db,
+                          {"user_id": USER_ID}, self.mock_collection)
+        # missing user_id
+        self.assertRaises(AttributeError, app.upsert_user_in_db,
+                          {"name": USER_NAME}, self.mock_collection)
+        # missing both
+        self.assertRaises(AttributeError, app.upsert_user_in_db,
+                          {}, self.mock_collection)
+        # no users should have been inserted
+        self.assertEqual(self.mock_collection.count_documents({}), 0)
+
+    def test_additional_info_still_inserted(self):
+        """A user_object with additional information should still be inserted.
+
+        Note: app.upsert_user_in_db should not insert the user if it is missing
+        the 'user_id' or 'name' attributes, but should insert the user if
+        the user has more attributes than those. If the user is malformatted,
+        app.upsert_user_in_db raises an AttributeError, else it upserts the
+        user and returns the new user object.
+        """
+        self.mock_collection = mongomock.MongoClient().db.collection
+        user_to_insert = {
+            "user_id": USER_ID,
+            "name": USER_NAME,
+            "additional_info": ADDITIONAL_INFORMATION
+        }
+        app.upsert_user_in_db(user_to_insert, self.mock_collection)
+        found_user = self.mock_collection.find_one({"user_id": USER_ID})
+        self.assertEqual(user_to_insert["user_id"], found_user["user_id"])
+        self.assertEqual(user_to_insert["name"], found_user["name"])
+        self.assertEqual(user_to_insert["additional_info"],
+                         found_user["additional_info"])
+
+    def test_multiple_upserts_is_one_insert(self):
+        """Upserting the same user multiple times should insert once."""
+        self.mock_collection = mongomock.MongoClient().db.collection
+        user_to_insert = {
+            "user_id": USER_ID,
+            "name": USER_NAME
+        }
+        # upsert many times
+        for i in range(0, 42):
+            app.upsert_user_in_db(user_to_insert, self.mock_collection)
+        # only 1 user has been inserted
+        self.assertEqual(self.mock_collection.count_documents({}), 1)
+        found_user = self.mock_collection.find_one({"user_id": USER_ID})
+        self.assertEqual(user_to_insert["user_id"], found_user["user_id"])
+        self.assertEqual(user_to_insert["name"], found_user["name"])
+
+    def test_returns_upserted_user(self):
```

Comment:
I would roll this test into the basic "happy path" in `test_insert_valid_user`.

Reply:
Done https://github.com/knative-portability/large-events/pull/8/commits/e88cb3cbf6afc45fc448a0d0eb76915db152da8d